src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs

```diff
@@ -369,22 +369,17 @@ public sealed override bool Equals(Symbol symbol, TypeCompareKind compareKind)
             if ((object)this == symbol) return true;
 
             return symbol is LambdaSymbol lambda
-                && areEqual(lambda.syntaxReferenceOpt, syntaxReferenceOpt)
+                && lambda.Syntax == Syntax
                 && lambda._refKind == _refKind
                 && TypeSymbol.Equals(lambda.ReturnType, this.ReturnType, compareKind)
                 && ParameterTypesWithAnnotations.SequenceEqual(lambda.ParameterTypesWithAnnotations, compareKind,
                                                                (p1, p2, compareKind) => p1.Equals(p2, compareKind))
                 && lambda.ContainingSymbol.Equals(ContainingSymbol, compareKind);
-
-            static bool areEqual(SyntaxReference a, SyntaxReference b)
-            {
-                return (object)a.SyntaxTree == b.SyntaxTree && a.Span == b.Span;
-            }
         }
 
         public override int GetHashCode()
         {
-            return syntaxReferenceOpt.GetHashCode();
+            return Syntax.GetHashCode();
```

Comment:
>Syntax.GetHashCode(); [](http://example.com/codeflow?start=19&length=21)

Is `Syntax.GetHashCode()` guaranteed to return the same value each time given that `Syntax` is implemented as `syntaxReferenceOpt.GetSyntax()`? 

If we want to use the `Syntax` for `GetHashCode()` and `Equals()`, then `LambdaSymbol` should probably hold onto the `SyntaxNode` in a `readonly` field instead. Otherwise `GetHashCode()` and `Equals()` should rely on the fields of `syntaxReferenceOpt` rather than `syntaxReferenceOpt.GetSyntax()`.

For the latter approach, perhaps:
```csharp
public override int GetHashCode()
{
    return Hash.Combine(
        syntaxReferenceOpt.SyntaxTree.GetHashCode(),
        syntaxReferenceOpt.Span.GetHashCode());
}
```

Reply:
@cston From the unit test, it seems to be working as expected. But I'm okay with either holding `Syntax` as a `_syntax` field, or implementing it using `SyntaxReference.SyntaxTree|Span`.